doc/dynamic-port-breakout/sonic-dynamic-port-breakout-HLD.md

```diff
@@ -0,0 +1,1182 @@
+
+Speeds in the breakout modes will be used for speed validation when we change the individual ports speed later.
+Eg, In `4x25G[10G]` mode, we can change the individual port in that port group to 25G or 10G, but not some other speeds.
+
+The speeds usually used today are:
+`10G, 20G, 25G, 50G, 100G, 200G, 400G` etc.
+However, the design is not limited to this list.
+
+We will have a new table `BREAKOUT_CFG` in configDB to present the current running breakout mode, it will be saved to configDB after the initial breakout at boot time, and also updated whenever the port breakout is changed.
+
+```
+BREAKOUT_CFG:
+{
+    "Ethernet0": {
+        "current_brkout_mode": "1x100G[40G]"
+    },
+
+    "Ethernet4": {
+        "current_brkout_mode": "2x50G"
+     }
+    ...
+}
+```
+
+The configDB will have the final port breakout settings for individual ports, the format/schema is the same as before. Above table is used for CLI to quickly know what breakout mode the system is on, and thus delete/add relevant ports if we are going to change the breakout mode. Also, it could be used by backend daemons like `PMON` to know the breakout mode and change the port layout for port based events and LED changes etc.
+
+After dynamic port breakout support, we won’t need different HWSKUs for the same platform. If for any reason (e,g, keep backwards compatibility) the user still wants to have HWSKUs, we would put the `platform.json` into each HWSKU with the desired breakout modes:
+```
+"Ethernet0": {
+    "index": "1,1,1,1",
+    "lanes": "0,1,2,3",
+    "alias_at_lanes": "Eth1/1, Eth1/2, Eth1/3, Eth1/4",
+    "breakout_modes": "1x100G[40G],2x50G,4x25G[10G],2x25G(2)+1x50G(2),1x50G(2)+2x25G(2)",
+    "default_brkout_mode": "1x100G[40G]"
+ },
+
+ "Ethernet4": {
+    "index": "2,2,2,2",
+    "lanes": "4,5,6,7",
+    "alias_at_lanes": "Eth2/1, Eth2/2, Eth2/3, Eth2/4",
+    "breakout_modes": "1x100G[40G],2x50G,4x25G[10G],2x25G(2)+1x50G(2),1x50G(2)+2x25G(2)",
+    "default_brkout_mode": "2x50G"
+ }
+…
+```
+The file will will have precedence over the `platform.json` at platform level, and it will be used for port breakout during the initialization phase
+
+Above `platform.json` will deprecate the old port_config.ini file in the current SONiC design.
+
+To keep backwards compatibility, we will modify the current code which is using port_config.ini to read from `platform.json` first, if it doesn't exist, it will read the port_config.ini as usual. To support the breakout feature, the json files will be required. In case they are missing, DPB changes would fail at CLI phase.
+
+### Special HW capability consideration
+On some platforms, there are some limitations about the number of port/mac available for the total count or for a particular group. For example, on some platforms, there were limitations where only 4 port/MAC are available for 8 lanes, in this case, we could define the `platform.json` to fit that platform. e,g:
+```
+"Ethernet0": {
```

Comment:
As we discussed with you in design review call, below are dell comments: 
1# What you are doing here is grouping the different ports in group and giving the name as starting interface name. We propose use consider interface name as group and derive interface names form context. 
Example: 
"Group0": {
    "index": "1,1,1,1,2,2,2,2",
    "lanes": "0,1,2,3,4,5,6,7",
  	"port_names": [
  					"Ethernet" : "Ethernet0, Ethernet4", 
  					"FibreChannel" : "FibreChannel0, FibreChannel4"
  					],
    "alias_at_lanes":"Eth1/1, Eth1/2, Eth1/3, Eth1/4, Eth2/1, Eth2/2, Eth2/3, Eth2/4",
    "breakout_modes": "2x100G[40G],4x50G,1x100G[40G](4)+2x50G(4),2x50G(4)+1x100G[40G](4),4x25G[10G](4), None(4)+4x25G[10G](4)",
    "default_mode": "Ethernet",
    "default_breakout_mode": "2x100G[40G]"
 }

2#  If you follow is above convention, it will have option to select port mode Ethernet/Fiber channel and derive default names used based on mode. it will be cleaner to isolate Ethernet/fiber channel ports in application. 

3# Mode can be changed dynamical, based on configured more Port name will be decided.
Example: 
"Group0": {
    "index": "1,1,1,1",
    "lanes": "0,1,2,3",
	  "port_names": {"Ethernet" : "Ethernet0", 
	                 "FibreChannel" : "FibreChannel0"
	                 },
    "alias_at_lanes": {"Ethernet" : "Eth1/1, Eth1/2, Eth1/3, Eth1/4",
                        "FibreChannel" : "FC1/1, FC1/2, FC1/3, FC1/4"},
    "breakout_modes": {"Ethernet" : "2x100G[40G],4x50G,1x100G[40G](4)+2x50G(4),2x50G(4)+1x100G[40G](4),4x25G[10G](4), None(4)+4x25G[10G](4)",
			            		 "FibreChannel" : "..." },
    "default_mode": "Ethernet",
    "default_brkout_mode": "2x100G[40G]"
}

Since we are doing this file format, we can consider this will be useful for future.

As per our discussion,  You will continue back end implementation and  update HLD will above proposal. 
Make necessary changes.